Add support for xpost suggestions on mobile

mchowning opened the pull request:
## Description
Adding xpost handling for mobile

Related:
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/2746

## How has this been tested?

In the demo app:
- Long-press the `@` toolbar button and choose Crosspost to insert a dummy Xpost.
- Long-press the `@` toolbar button and choose Mention to check that a mention is still inserted
- Tap the `@` button to make sure that a mention is still inserted


## Screenshots <!-- if applicable -->

| Android | iOS |
| - | - |
| <img src="https://user-images.githubusercontent.com/1898325/103029176-fa500c80-4537-11eb-8d9f-cc5473933e04.gif" alt="Android screen recording" width="350"> | <img src="https://user-images.githubusercontent.com/1898325/103028865-5b2b1500-4537-11eb-9ed5-2a51a7579ec6.gif" alt="iOS screen recording" width="350">  |

## Types of changes
- New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

mchowning committed:
Add support for xpost suggestions on mobile

Commit:
Add xpost bridge methods for iOS

Commit:
Update @mention toolbar button callback

Update it to the new callback function

mchowning committed:
Check for trigger key codes at beginning of lines

mchowning committed:
Add xposts capability handling

This is limited to Android at this time

Commit:
Merge branch 'master' into rnmobile/xpost_initial_implementation

Commit:
added xposts capability for iOS

Commit:
Merge branch 'master' into rnmobile/xpost_initial_implementation

Commit:
Add Xpost to toolbar via long-press on @-mentions

Long-pressing the @-mentions toolbar button will now let the user add an Xpost or an @-mention

guarani commented on packages/rich-text/src/component/index.native.js:
```diff
@@ -961,13 +981,32 @@ export class RichText extends Component {
 						</BlockFormatControls>
 					</>
 				) }
+				<Picker
+					ref={ ( instance ) => ( this.suggestionPicker = instance ) }
+					options={ [
+						{
+							value: 'mention',
+							label: __( 'Mention' ),
+							icon: atSymbol,
+						},{
+							value: 'xpost',
+							label: __( 'Xpost' ),
+							icon: plus,
+						}
+					] }
+					onChange={ onSuggestionPickerSelect }
+					hideCancelButton
+				/>
```
@mchowning I wasn't sure where to add this `Picker` since it's rendered outside of the component hierarchy, but it seems like anywhere will do so I chose here. Let me know if you think it belongs somewhere else.

Also, this file is huge and I've added more lines of code. Not sure if we should try to move this to a new file, but left it as-is for now.

Commit:
Updated code formatting

mchowning committed:
Merge remote-tracking branch 'origin/master' into rnmobile/xpost_initial_implementation

mchowning committed:
Dynamically update suggestions button

mchowning committed:
Update snapshot tests broken by xpost modal

The addition of the xposts modal in
e730f2e6c45ce8ddb33ffae10babefd5c16cb530 broke these snapshot tests.

mchowning committed:
Extract suggestions format button component

mchowning committed:
Make suggestions toolbar button generic

mchowning committed:
Replace Toolbar with ToolbarGroup

Because using Toolbar with a label prop is deprecated: https://github.com/WordPress/gutenberg/blob/af768dc98c625e33de8023084c6a5c198c52f8ef/packages/components/src/toolbar/index.js#L30-L34

mchowning committed:
Update mobile snapshot tests to remove Modal

cameronvoell commented on packages/react-native-editor/ios/GutenbergDemo/GutenbergViewController.swift:
```diff
@@ -225,6 +225,10 @@ extension GutenbergViewController: GutenbergBridgeDelegate {
     }
 
+    func gutenbergDidRequestXpost(callback: @escaping (Result<String, NSError>) -> Void) {
+        callback(.success("ma.tt"))
```
same as https://github.com/WordPress/gutenberg/pull/26211/files#r540769080, ios demo app didnt seem to have "mocked" cross posts working, perhaps this is just a placeholder.

mchowning replied:
Thanks! Addressed on iOS in bb6ecb5. 

I'm not very familiar with the setup of the iOS demo app, so @guarani , if you would take a look at that commit and make sure there's nothing else I needed to do, I would appreciate it.

guarani replied:
I think that looks good, thanks for updating that @mchowning!

mchowning committed:
Enable xposts in Android demo app

mchowning committed:
Enable xposts in ios demo app

mchowning commented on packages/react-native-editor/android/app/src/main/java/com/gutenberg/MainApplication.java:
```diff
@@ -189,8 +189,13 @@ public void gutenbergDidRequestUnsupportedBlockFallback(ReplaceUnsupportedBlockC
+
+            @Override
+            public void onShowXpostSuggestions(Consumer<String> onResult) {
+                onResult.accept("ma.tt");
```
Good catch. We added the placeholder data, but forgot turn the capability on in the demo app. This should be working now on both platforms.

mchowning commented:
👋 @jhnstn! If you have a chance, I would appreciate it if you could take a look at the js code in this PR.

Commit:
Add change log item for cross-posts

Commit:
Merge branch 'master' into rnmobile/xpost_initial_implementation

cameronvoell reviewed (approved):
Demo app looks great, as well as the rest of the changes here. :shipit:

hypest commented:
👋 @mchowning and @guarani , please add some testing steps (here or in the gutenberg-mobile PR), thanks!

Commit:
Merge branch 'master' into rnmobile/xpost_initial_implementation

guarani reviewed (approved):
Working great from my perspective! Tested demo apps on both platforms, I think this is ready to merge.

Thanks @mchowning!

jhnstn commented on packages/rich-text/src/component/index.native.js:
```diff
@@ -401,33 +412,66 @@ export class RichText extends Component {
+				icon: plus,
+			},
+		];
+		return allOptions.filter( ( op ) => op.supported );
```
I personally prefer using _functional_ flows like using `filter` here but I think it might be more familiar to build `allOptions` with  conditionals :
```
const allOptions = [];
if ( areMentionsSupported ) {
  allOptions.push({<option object>})
}

if ( areXPostsSupported ) {
  allOptions.push({<option object>})
}

return allOptions;
``

jhnstn commented on packages/rich-text/src/component/index.native.js:
```diff
@@ -401,33 +412,66 @@ export class RichText extends Component {
-	handleMention( event ) {
+	handleTriggerKeyCodes( event ) {
 		const { keyCode } = event;
+		const triggeredOption = this.suggestionOptions().find( ( option ) => {
```
Can the return from `this.suggestionOptions` be memorized to a class property so doesn't have to be called on every  key  code event?

jhnstn commented:
@mchowning  looks good! I left a few comments but nothing blocking.
